packages/core/src/lib/connection_manager.ts

```diff
@@ -164,21 +217,50 @@ export class ConnectionManager {
     );
   }
 
+  private async attemptDial(peerId: PeerId): Promise<void> {
+    if (this.currentActiveDialCount >= this.options.maxParallelDials) {
+      this.pendingPeerDialQueue.push(peerId);
+      return;
+    }
+
+    if (!(await this.shouldDialPeer(peerId))) return;
+
+    this.dialPeer(peerId).catch((err) =>
+      log(`Error dialing peer ${peerId.toString()} : ${err}`)
+    );
+  }
+
   private onEventHandlers = {
     "peer:discovery": async (evt: CustomEvent<PeerInfo>): Promise<void> => {
       const { id: peerId } = evt.detail;
-      if (!(await this.shouldDialPeer(peerId))) return;
 
-      this.dialPeer(peerId).catch((err) =>
+      this.attemptDial(peerId).catch((err) =>
         log(`Error dialing peer ${peerId.toString()} : ${err}`)
       );
     },
-    "peer:connect": (evt: CustomEvent<Connection>): void => {
-      {
-        this.keepAliveManager.start(
-          evt.detail.remotePeer,
-          this.libp2pComponents.ping.bind(this)
-        );
+    "peer:connect": async (evt: CustomEvent<Connection>): Promise<void> => {
+      const { remotePeer: peerId } = evt.detail;
+
+      this.keepAliveManager.start(
+        peerId,
+        this.libp2pComponents.ping.bind(this)
+      );
+
+      const isBootstrap = (await this.getTagNamesForPeer(peerId)).includes(
+        Tags.BOOTSTRAP
+      );
+
+      if (isBootstrap) {
+        const bootstrapConnections = this.libp2pComponents
```

Comment:
Is it possible to determine `isBootstrap` by checking `conn.tags`? If so we can make `shouldDialPeer` sync and create a method to remove duplication of
```ts
this.libp2pComponents
        .getConnections()
        .filter((conn) => {
          conn.tags.find((name) => name === Tags.BOOTSTRAP);
        })```

Reply:
Then lest resolve that comment in another PR or as follow-up